keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,884 @@
+  3. The request has a header that indicates that this request has been proxied once already. If for some reason the resource is not found locally, we will serve error 503
+  4. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+  5. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 404 with error `GVR <group_version_resource> is not served by anything in this cluster`
+
+5. If the proxy call fails for network issues or any reason, we serve 503 with error `Error while proxying request to destination apiserver`
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+##### StorageVersion enhancement needed
+
+StorageVersion API currently tells us whether a particular StorageVersion can be read from etcd by the listed apiserver. We will enhance this API to also include apiserver ID of the server that can serve this StoageVersion.
+
+#### Identifying destination apiserver's network location
+
+* TODO: We need to find a place to store and retrieve the destination apiserver's host and port information given the server's ID.
+We do not want to store this information in
+
+    * StorageVersion : because we do not want to expose the network identity of the apiservers in this API that can be listed in multiple places where it may be unnecessary/redundant to do so
+    * Endpoint reconciler lease : because the IP present here could be that of a load balancer for the apiservers, but we need to know the definite address of the identified destination apiserver
+
+#### Proxy transport between apiservers and authn
+
+For the mTLS between source and destination apiservers, we will do the following
+
+1. For server authentication by the client (source apiserver) : the client needs to validate the server certs (presented by the destination apiserver), for which it needs to know the CA bundle of the authority that signed those certs. We should be able to reuse the bundle given to all pods to verify whatever kube-apiserver instance they talk to (currently passed to kube-controller-manager as --root-ca-file)
+
+2. For client authentication by the server (destination apiserver) : destination apiserver will check the source apiserver certs to determine that the proxy request is from an authenticated client. The destination apiserver will use requestheader authentication (and NOT client cert authentication) for this using the kube-aggregator proxy client cert/key and the --requestheader-client-ca-file passed to the apiserver upon bootstrap
```

Comment:
etcd has configuration for both client-to-server and peer-to-peer traffic. Long term, I expect that we'll need to converge to something similar, only we also have kubelet traffic, so there actually three "directions" or traffic.

Can we model this after etcd (https://etcd.io/docs/v3.2/op-guide/security/)?  I'm OK with the peer-to-peer traffic defaulting to the approach defined here, but I think we should also offer optional flags to configure this new direction of traffic (e.g. `--peer-advertise-address`, `--peer-bind-address`, `--peer-client-ca-file` ,...).

cc @deads2k